docs/adr/0012-runtime-message-results.md

```diff
@@ -0,0 +1,167 @@
+# ADR 0012: Runtime Message Results
+
+## Changelog
+
+- 2021-12-04: Initial version
+
+- 2021-12-10: Extend the implementation section
+
+## Status
+
+Proposed
+
+## Context
+
+Currently, the results of emitted runtime messages are `MessageEvent`s, which
+only provide information whether the message execution was successful or not. For
+various use-cases additional information about message results would be useful.
+
+One of such is supporting staking by runtimes. Currently, a runtime can emit an
+`AddEscrow` message, but is unaware of the actual amount of shares it obtained
+as a result of the added escrow. For some use-cases (e.g. runtime staking user
+deposited funds) this information is crucial for accounting.
+
+Similarly, for `ReclaimEscrow`, the runtime doesn't have the direct information
+at which epoch the stake gets debonded.
+
+The only way to currently obtain this data is to subscribe to consensus events,
+something which runtime doesn't have access to.
+
+Adding results to `MessageEvent` solves both of the mentioned use cases:
+
+- for `AddEscrow` the result should contain amount of shares obtained with the
+  escrow
+
+- for `ReclaimEscrow` the result should contain the amount of shares and epoch
+  at which the stake gets debonded
+
+## Decision
+
+Implement support for arbitrary result data in `MessageEvent` runtime message
+results.
+
+## Implementation
+
+- Result field is added to `roothash.MessageEvent` struct:
+
+```golang
+// MessageEvent is a runtime message processed event.
```

Comment:
Are we going to impose a sensible limit on the size of each message, and the number of messages a runtime can publish at once?

Reply:
This is already the case with the existing `MaxMessages` configuration option. There is no explicit limit on size of an individual message, but there is an upper limit which is the maximum transaction size. The messages however cannot be arbitrary, they are limited to specific actions (currently staking operations and runtime descriptor updates).

Reply:
Ok.